blueprints/layered-profile.md

```diff
@@ -0,0 +1,274 @@
+    ],
+    "profileNames": [
+      "MID", 
+      "TOP", 
+      "PRIMARY"
+    ],
+    ⋮
+  },
+  ⋮
+}
+```
+
+The following table describes the top level `layered_profile` object for delivery services:
+
+| field              | type          | optionality | description                                                    |
+| ------------------ | --------------| ----------- | ---------------------------------------------------------------|
+| deliveryservice    | bigint        | required    | the delivery service id associated with a given profile        |
+| profile            | bigint        | required    | the profile id associated with a delivery service              |
+| order              | bigint        | required    | the order in which a profile is applied to a delivery service  |
+| lastUpdated        | bigint        | required    | the last time this delivery service was updated                |
+
+The following table describes the top level `layered_profile` object for servers:
+
+| field       | type                 | optionality | description                                                    |
+| ----------- | ---------------------| ----------- | ---------------------------------------------------------------|
+| server      | bigint               | required    | the server id associated with a given profile                  |
+| profile     | bigint               | required    | the profile id associated with a server                        |
+| order       | bigint               | required    | the order in which a profile is applied to a server            |
+| lastUpdated | bigint               | required    | the last time this server was updated                          |
+
+**API constraints**
+- TO REST APIs will not be backward compatibility and this feature will be a major version (v5.0) change.
+- Add `Profiles` key to API endpoints for Server and DeliveryService objects (and ProfileNames, ProfileIDs)
+
+#### Client Impact
+- Existing Go client methods will be updated for the `/deliveryservices` and `/servers` endpoints in order to write TO API tests for the exising endpoints.
+- All functions which get Parameters on Delivery Services or Servers must be changed to get the Parameters of all assigned Profiles in order.
+
+#### Data Model / Database Impact
+
+- A new Database Table `server_profiles` as described below, will be created:
+- A new Database Table `deliveryservice_profiles` as described below, will be created:
+```text
+            Table "traffic_ops.deliveryservice_profiles"
+     Column     |  Type                    | Collation | Nullable | Default
+----------------+--------------------------+-----------+----------+--------
+deliveryservice | bigint                   |           | not null |
+profile         | bigint                   |           | not null |
+order           | bigint                   |           | not null |
+last_updated    | timestamp with time zone |           | not null | now()
```

Comment:
Does this actually need a `last_updated`? When would that be used or displayed? Could we instead just include a trigger to change the `last_updated` of affected Delivery Services when changes are made in this table?

Reply:
We have last updated on pretty much every joining table like this (`topology_cachegroup_parents`, `server_server_capability`, `federation_federation_resolver`, etc.).  For consistency I'd think we should keep `last_updated`.

I don't actually know what we use them for but I imagine for "auditing"/debugging

Reply:
That will increase the complexity of checking for request preconditions. We also have join tables (actually just one, that I know of) that use triggers instead of their own `last_updated`, which I think is worth using moving forward because of how easy it makes things.

Reply:
+1 on a trigger to just update `last_updated` on the "main" table

Reply:
I just saw the prior two comments. I think having `last_updated` on the join tables as well as the main tables would be necessary if we have APIs that just return the "join" data? E.g. if we had an API that returned all the server_server_capability data, it would make sense to have a last_updated on that. 

For _this_ feature, I'm not sure it would even be necessary to add a trigger though, because a PUT is going to update all the server fields anyways, including the new profile list, which will trigger the last_updated trigger on the server table.

Reply:
That's true, our current implementation runs update queries even if nothing changes - because it doesn't check for changes. It's probably unnecessary unless postgres makes some kind of optimizations that would somehow skip unchanging updates and the associated triggers (which would surprise me). I wouldn't be opposed to adding the trigger anyway, but given that our strategy for updates isn't likely to change in this way any time soon, I agree the trigger sounds pretty optional.